[libc++][math] Provide overloads for cv-unqualified floating point types for `std::signbit`

robincaloudis opened the pull request:
## Why
Following up on https://github.com/llvm/llvm-project/pull/105946, this patch provides the floating point overloads for `std::signbit` as defined by [P0533R9](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0533r9.pdf).

## What
* Test and add overloads for cv-unqualified floating point types
* Remove constrained overload as it is not needed anymore
* Make use of `template<class = void>` as the universal C runtime (UCRT) needed for Clang-Cl comes with overloads for all cv-unqualified floating point types (float, double, long double) for `std::signbit()` by itself [in the WinSDK](https://github.com/microsoft/win32metadata/blob/e012b29924c53aa941fc010850b68331b0c3ea80/generation/WinSDK/RecompiledIdlHeaders/ucrt/corecrt_math.h#L309-L322). In a certain way, this can be seen as a deviation from the C standard. We need to work around it as the compilation would otherwise error out due to duplicated definitions.

Zingam commented on libcxx/docs/Status/Cxx23Papers.csv:
```diff
@@ -39,7 +39,7 @@
 "","","","","",""
 "`P0323R12 <https://wg21.link/P0323R12>`__","``std::expected``","2022-02 (Virtual)","|Complete|","16.0",""
-"`P0533R9 <https://wg21.link/P0533R9>`__","``constexpr`` for ``<cmath>`` and ``<cstdlib>``","2022-02 (Virtual)","|In Progress|","","``isfinite``, ``isinf``, ``isnan`` and ``isnormal`` are implemented"
+"`P0533R9 <https://wg21.link/P0533R9>`__","``constexpr`` for ``<cmath>`` and ``<cstdlib>``","2022-02 (Virtual)","|In Progress|","","``isfinite``, ``isinf``, ``isnan``, ``isnormal`` and ``signbit`` are implemented"
```
Nit: IMO it would be better to track the progress as subtasks in https://github.com/llvm/llvm-project/issues/105174

robincaloudis replied:
I was not aware that we have GitHub issues for each paper/issue/DR that are fully synchronized with the CSV status files. I will track the progress as subtask. Thanks for the friendly reminder!

Zingam replied:
@robincaloudis Please let me know if you need help with creating the subtasks while you wait for commit access.

robincaloudis replied:
Thanks. I will do so.

robincaloudis committed:
Test floating point overloads

robincaloudis committed:
Add floating point overloads for std::signbit

By using `_LIBCPP_PREFERRED_OVERLOAD` we make
sure that a given overload is a better match
than an otherwise equally good function
declaration.

Why is there an equally good function
declaration in the first place? Underlying the
Windows SDK is the UCRT, the universal C runtime,
which clang-cl makes use of. The UCRT should
provide only C library headers, but does on top
comes with overloads for all cv-unqualified floating
point types (float, double, long double) for
`std::signbit()` in https://github.com/microsoft/win32metadata/blob/e012b29924c53aa941fc010850b68331b0c3ea80/generation/WinSDK/RecompiledIdlHeaders/ucrt/corecrt_math.h#L309-L322.
In a certain way, this can be seen as a deviation
from the C standard. We need to work around it.

philnik777 commented on libcxx/include/__math/traits.h:
```diff
@@ -34,8 +34,30 @@ namespace __math {
-template <class _A1, __enable_if_t<is_floating_point<_A1>::value, int> = 0>
-_LIBCPP_NODISCARD inline _LIBCPP_SIGNBIT_CONSTEXPR _LIBCPP_HIDE_FROM_ABI bool signbit(_A1 __x) _NOEXCEPT {
+_LIBCPP_NODISCARD inline _LIBCPP_SIGNBIT_CONSTEXPR _LIBCPP_HIDE_FROM_ABI
+#ifdef _LIBCPP_PREFERRED_OVERLOAD
+_LIBCPP_PREFERRED_OVERLOAD
+#endif
```
Why can't we instead make this a template like we do for other functions which may have already been provided by the libc?

robincaloudis replied:
Good idea. I made a proposal. What do you think?

philnik777 replied:
I meant that we can instead add `template <class = void>` to make the overloads weaker than the UCRT overloads. Is there any reason not to do that?

robincaloudis replied:
Now I understand what you meant. Done.

robincaloudis committed:
Encapsulate the use of macro and leave comment

robincaloudis committed:
Define overloads as templates

philnik777 commented on libcxx/include/__math/traits.h:
```diff
@@ -34,8 +34,21 @@ namespace __math {
 
-template <class _A1, __enable_if_t<is_floating_point<_A1>::value, int> = 0>
-_LIBCPP_NODISCARD inline _LIBCPP_SIGNBIT_CONSTEXPR _LIBCPP_HIDE_FROM_ABI bool signbit(_A1 __x) _NOEXCEPT {
+// The universal C runtime (UCRT) in the WinSDK provides floating point overloads
```
I think this removes the only use of `is_floating_point`.

robincaloudis replied:
You are right. I removed the header.

robincaloudis committed:
Remove unused header
